### PR TITLE
net: lib: Erase pending may cause the modem FOTA to hang 

### PR DESCRIPTION
**Pull request by sigvartmh**
Doing a modem DFU may cause the modem to never execute the erase just functioning normally so the code will hang on `DFU_ERASE_PENDING`.
This is due to the modem being too busy doing normal operations than to do the actual erase. Either
reboot the device or turn of the LTE link until the erase operation is
complete to mitigate this issue.

### CODE REVIEW

**Comment by sigvartmh** [comment]
This is not complete just made it a draft PR for now. It has to be implemented for other applications and samples also.

File: subsys/dfu/src/dfu_target_modem.c [hunk 1]
```diff
@@ -122,12 +130,14 @@ bool dfu_target_modem_identify(const void *const buf)
...
 
-int dfu_target_modem_init(size_t file_size)
+int dfu_target_modem_init(size_t file_size, dfu_target_callback_t cb)
 {
 	/* We have no way of checking the amount of flash available in the modem
 	 * with the current API
 	 */
 	ARG_UNUSED(file_size);
+	callback = cb;
+
```
Comment (sigvartmh) [inline subsys/dfu/src/dfu_target_modem.c]: excessive new line

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 #include <zephyr.h>
 #include <stdio.h>
+#include <lte_lc.h>
```
Comment (sigvartmh) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c]: Not needed if you won't be doing a `lte_lc_offline`/`lte_lc_connect`.

What I found is that you get a 
```
[00:00:55.537,261] <err> asset_tracker: Socket error: POLLHUP
[00:00:55.543,975] <err> asset_tracker: Connection was closed by the cloud.
[00:00:55.551,818] <err> asset_tracker: LTE link disconnect
``` 
If you do it so I guess some more state handling/shifting is needed.

File: subsys/dfu/src/dfu_target_modem.c [hunk 3]
```diff
@@ -56,25 +58,31 @@ static int delete_banked_modem_fw(void)
...
 		LOG_ERR("Failed to delete backup, errno %d", errno);
 		return -EFAULT;
 	}
-
 	while (true) {
 		err = getsockopt(fd, SOL_DFU, SO_DFU_OFFSET, &offset, &len);
 		if (err < 0) {
+			if (timeout < 0) {
+				callback(DFU_TARGET_EVT_TIMEOUT);
+				timeout = 60;
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/dfu/src/dfu_target_modem.c]: I'll add it as a Kconfig option.

File: subsys/dfu/src/dfu_target_modem.c [hunk 4]
```diff
@@ -122,12 +130,14 @@ bool dfu_target_modem_identify(const void *const buf)
 
 }
 
-int dfu_target_modem_init(size_t file_size)
+int dfu_target_modem_init(size_t file_size, dfu_target_callback_t cb)
 {
 	/* We have no way of checking the amount of flash available in the modem
 	 * with the current API
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply subsys/dfu/src/dfu_target_modem.c]: Older version of the modem/bsdlib would just return the flash size of the modem. Is this now fixed in bsdlib/modem?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c (hunk 2 above)
Comment (sigvartmh) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c]: After consulting those who are responsible for `asset_tracker` it was decided that a `lte_lc_power_off`would be the best solution

File: samples/nrf9160/aws_fota/src/main.c [hunk 5]
```diff
@@ -438,32 +441,43 @@ static int fds_init(struct mqtt_client *c)
  *
  * Blocks until link is successfully established.
```
Comment (jorgenmk) [inline samples/nrf9160/aws_fota/src/main.c]: The doxygen for this function is no longer valid

**Comment by thst-nordic** [comment]
hold of testing for a little bit.  this is going to need a rebase to fix the failing documentation build.

File: samples/nrf9160/aws_fota/src/main.c (hunk 5 above)
Comment (sigvartmh) [reply samples/nrf9160/aws_fota/src/main.c]: Fixed.

**Comment by thst-nordic** [comment]
master is ready for rebase to fix documentation build.

**Comment by carlescufi** [comment]
@sigvartmh @thst-nordic is this an actual CI issue?

**Comment by sigvartmh** [comment]
@carlescufi  It's an actual CI failure I think I've fixed it now but we'll have to wait until CI is green to see if I was successful.

File: samples/nrf9160/aws_fota/src/main.c [hunk 6]
```diff
@@ -434,36 +437,44 @@ static int fds_init(struct mqtt_client *c)
...
 		printk("AWS_FOTA_EVT_DONE, rebooting to apply update.\n");
 		do_reboot = true;
 		break;
 
+	case AWS_FOTA_EVT_ERASE_PENDING:
+		printk("AWS_FOTA_EVT_ERASE_PENDING, reboot or disconnect the "
+		       "LTE link\n");
+		err = lte_lc_offline();
+		if (err) {
+			printf("Error turning off the LTE link\n");
```
Comment (lemrey) [inline samples/nrf9160/aws_fota/src/main.c]: `printf` -> `printk` here and line 474.

File: subsys/dfu/src/dfu_target_modem.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (lemrey) [reply subsys/dfu/src/dfu_target_modem.c]: > Older version of the modem/bsdlib would just return the flash size of the modem

I don't think this has ever been the case for `NRF_SO_DFU_RESOURCES`

File: samples/nrf9160/aws_fota/src/main.c (hunk 6 above)
Comment (sigvartmh) [reply samples/nrf9160/aws_fota/src/main.c]: Thanks for spotting it was an autocomplete error :+1:

**Comment by sigvartmh** [comment]
@lemrey Updated commit message to mention DFU operation so it better reflects what is hanging.

**Comment by carlescufi** [comment]
@lemrey please revisit

**Pull request merged**